src/mapgen.cpp

```diff
@@ -303,42 +409,27 @@ load_mapgen_function( const JsonObject &jio, const std::string &id_base,
             if( jio.has_string( "name" ) ) {
                 const std::string mgname = jio.get_string( "name" );
                 if( mgname == id_base ) {
-                    oter_mapgen[id_base][ default_idx ]->weight = 0;
+                    oter_mapgen.erase( id_base, default_idx );
                 }
             }
         }
         jio.allow_omitted_members();
         return nullptr; // nothing
-    } else if( jio.has_string( "method" ) ) {
-        const std::string mgtype = jio.get_string( "method" );
-        if( mgtype == "builtin" ) { // c-function
-            if( jio.has_string( "name" ) ) {
-                const std::string mgname = jio.get_string( "name" );
-                if( const auto ptr = get_mapgen_cfunction( mgname ) ) {
-                    ret = std::make_shared<mapgen_function_builtin>( ptr, mgweight );
-                    oter_mapgen[id_base].push_back( ret );
-                } else {
-                    debugmsg( "oter_t[%s]: builtin mapgen function \"%s\" does not exist.", id_base.c_str(),
-                              mgname );
-                }
-            } else {
-                debugmsg( "oter_t[%s]: Invalid mapgen function (missing \"name\" value).", id_base.c_str() );
-            }
-        } else if( mgtype == "json" ) {
-            if( jio.has_object( "object" ) ) {
-                JsonObject jo = jio.get_object( "object" );
-                std::string jstr = jo.str();
-                ret = std::make_shared<mapgen_function_json>( jstr, mgweight, offset );
-                oter_mapgen[id_base].push_back( ret );
-            } else {
-                debugmsg( "oter_t[%s]: Invalid mapgen function (missing \"object\" object)", id_base.c_str() );
-            }
+    }
+    const std::string mgtype = jio.get_string( "method" );
+    if( mgtype == "builtin" ) {
+        if( const auto ptr = get_mapgen_cfunction( jio.get_string( "name" ) ) ) {
+            ret = std::make_shared<mapgen_function_builtin>( ptr, mgweight );
+            oter_mapgen.add( id_base, ret );
```

Comment:
The function says that it would `check_mapgen_definitions`, but it also does something else: mutates the internal state. I see that it was written that way before, but since it's heavily touched now, could that be fixed?

Reply:
I don't get it. *What* function says something? `oter_mapgen.add`? As far as I can see, nothing in this file mentions "check_mapgen_definitions", there is only the function definition itself.

Reply:
Oh, I misread the diff: it's actually `load_mapgen_function`, not `check_mapgen_definitions`. Please, disregard this comment.